New docs and examples (#84)
3scale-labs/authorino: docs/deploy.md

# Deploying Authorino

1. [Clone the repo](#clone-the-repo)
2. [Meet the requirements](#meet-the-requirements)
3. [Deploy](#deploy)
   - [Option A: Local cluster](#option-a-local-cluster)
   - [Option B: Custom deployment](#option-b-custom-deployment)

## Clone the repo

This is the easiest way to get Authorino's Kubernetes resources and deployment utils.

```sh
git clone git@github.com:3scale-labs/authorino.git && cd authorino
```

## Meet the requirements

Having a [Kubernetes](https://kubernetes.io/) server up and running is the only actual requirement to deploy Authorino.

> With the tools provided in the Authorino repo nevertheless, you can easily have a local cluster setup (using [Kind](https://kind.sigs.k8s.io)). In this case, the only requirement is [Docker](https://docker.com).

Reviewer: as in, no need for kubernetes?

Reviewer: => Nevertheless, with the tools provided in the Authorino repo…

Author: Correct. Kind starts a Kubernetes cluster in a docker container.

> ### Requirements: the untold truth

Reviewer: ```suggestion
### Feature-specific requirements
```

> Apart from the obvious requirement of having a running instance of a Kubernetes server, Authorino also rely on some other components and capabilities to perform its function. A brief list of such indispensable parts follows:

Reviewer: ```suggestion
Apart from the obvious requirement of having a running instance of a Kubernetes server, Authorino also relies on some other components and capabilities to perform its function:
```

<!-- TODO: Add minimum required Kubernetes version -->

> - Permission over the Kubernetes server to create Custom Resource Definitions (CRDs). This is actually a one-time requisite, necessary only to install Authorino, and assumed to be known by cluster administrators who should be able to handle as desired through the Kubernetes API Role-Based Access Control bindings.

Reviewer: ```suggestion
- Permission from the Kubernetes server to create Custom Resource Definitions (CRDs) during Authorino's installation. Cluster administrators can handle this requirement through the Kubernetes API Role-Based Access Control bindings.
```

Reviewer: Actually i would consider skipping this whole paragraph as it's true for anything that needs to create CRDs. Or maybe add it to the first part where you talk about k8s.

Author: I will keep it (moving up as you suggested). I think reminding about having privileges to create CRDs as a requirement won't harm.

> - [Envoy](https://www.envoyproxy.io) proxy (or, technically, any proxy that implements the client-side of the [gRPC protocol](https://www.envoyproxy.io/docs/envoy/latest/start/sandboxes/ext_authz). With Envoy, ultimately, virtual hosts will be associated to the upstream APIs to protect, and an [External Authorization Filter](https://www.envoyproxy.io/docs/envoy/latest/start/sandboxes/ext_authz) should point to Authorino. Check [this example](/examples/envoy/envoy-deploy.yaml) if you need to.

Reviewer: I would add this paragraph to the first part too and then rename the subtitle as suggested.

- For UMA-protected resource data, you will need a UMA-compliant server running as well. This can be an implementation of the UMA protocol by each upstream API itself or (more tipically) an external server that knows about the resources. Again, Keycloak can be a good fit here as well. Just keep in mind that, whatever resource server you choose, changing-state actions commanded in the upstream APIs or other parties will have to be reflected in the resource server. Authorino will not do that for you.
- For Kubernetes authentication tokens, in case you want to be able to requests access tokens for clients running outside the custer, you may want to check out the requisites for using Kubernetes [TokenRquest API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#tokenrequest-v1-storage-k8s-io) (GA in v1.20).

## Deploy

Choose between the options below to continue:

[**Option A:** Local cluster (try it out and examples)](#option-a-local-cluster)<br/>
[**Option B:** Custom deployment](#option-b-custom-deployment)

### Option A: Local cluster

> Option A is essentially meant for trying out Authorino locally. It brings a bundle of a [Kind](https://kind.sigs.k8s.io)-managed Kubernetes cluster, with a freshly built Authorino image loaded up and pre-configured sample resources to make testing and trying out simple and easy.

Reviewer: ```suggestion
Option A is meant for trying out Authorino locally. It gives you a bundle of a [Kind](https://kind.sigs.k8s.io)-managed Kubernetes cluster, with a freshly built Authorino image and pre-configured sample resources. 
```
